neuralforecast/common/_base_recurrent.py

```diff
@@ -103,28 +100,45 @@ def configure_optimizers(self):
         return torch.optim.Adam(self.parameters(), lr=self.learning_rate)
 
     def _normalization(self, batch):
-        temporal = batch['temporal']
-        temporal_cols = batch['temporal_cols']
+
+        temporal = batch['temporal'] # B, C, T
+        temporal_cols = batch['temporal_cols'].copy()
+
+        # Separate data and mask
+        temporal_data_cols = temporal_cols.drop('available_mask').tolist()
+        temporal_data = temporal[:, temporal_cols.get_indexer(temporal_data_cols), :]
+        temporal_mask = temporal[:, temporal_cols.get_loc('available_mask'), :].clone()
 
         # Remove validation and test set to prevent leakeage
         if self.val_size + self.test_size > 0:
-                cutoff = -self.val_size - self.test_size
-                temporal = temporal[:, :, :cutoff]
-
-        temporal_y = temporal[:, temporal_cols.get_loc('y'), :]
-        temporal_mask = temporal[:, temporal_cols.get_loc('available_mask'), :]
+            cutoff = self.val_size + self.test_size
+            temporal_mask[:, -cutoff:] = 0
 
-        y_means = torch.sum(temporal_y * temporal_mask, dim=-1, keepdim=True) / torch.sum(temporal_mask, dim=-1, keepdim=True)
-        y_stds = torch.sqrt(torch.sum(temporal_mask*(temporal_y-y_means)**2, dim=-1, keepdim=True)/ \
-                            torch.sum(temporal_mask, dim=-1, keepdim=True) )
+        # Normalize. self.scaler stores the shift and scale for inverse transform
+        temporal_mask = temporal_mask.unsqueeze(1) # Add channel dimension for scaler.transform.
+        temporal_data = self.scaler.transform(x=temporal_data, mask=temporal_mask)
 
-        temporal_y = (temporal_y - y_means) / y_stds
-        temporal[:, temporal_cols.get_loc('y'), :] = temporal_y
+        # Replace values in windows dict
+        temporal[:, temporal_cols.get_indexer(temporal_data_cols), :] = temporal_data
+        batch['temporal'] = temporal
         
-        return batch, y_means, y_stds
+        return batch
 
-    def _inv_normalization(self, y_hat, y_means, y_stds):
-        return y_stds[:, None]*y_hat + y_means[:, None]
+    def _inv_normalization(self, y_hat, temporal_cols):
+        # Receives window predictions [B, W, H]
+        # Broadcasts outputs and inverts normalization
+
+        # Get y scale and shift, and add W dimension
+        temporal_data_cols = temporal_cols.drop('available_mask')
+        y_scale = self.scaler.x_scale[:, temporal_data_cols.get_indexer(['y']), :]
+        y_shift = self.scaler.x_shift[:, temporal_data_cols.get_indexer(['y']), :]
+
+        y_scale = torch.repeat_interleave(y_scale, repeats=y_hat.shape[1], dim=1)
+        y_shift = torch.repeat_interleave(y_shift, repeats=y_hat.shape[1], dim=1)
+
+        y_hat = self.scaler.inverse_transform(z=y_hat, x_scale=y_scale, x_shift=y_shift)
```

Comment:
Is this transformation respecting the convention of the order for the `y_hat` dimensions?
\# [N,T,C,H,Q]?

This might be problematic.

Reply:
yes